### ISSUE TITLE
Restrict fixed effects parameters when using bounded priors

### ISSUE THREAD

**Issue by joelgombin** [paul-buerkner/brms#29]
I don't manage to use uniform priors in `brm`. Consider this (admittedly quite silly) example. Is there something I missing or is that a bug?

```r
library(brms)
data("cars")
lm(dist ~ speed, data = cars)
### Call:
### lm(formula = dist ~ speed, data = cars)
###
### Coefficients:
### (Intercept)        speed  
###     -17.579        3.932 

test1 <- brm(dist ~ speed, data = cars, family = gaussian(link = "identity"), prior = set_prior("normal(0, 10)", class = "b")) # works
test2 <- brm(dist ~ speed, data = cars, family = gaussian(link = "identity"), prior = set_prior("uniform(2, 4)", class = "b", coef = "speed")) # doesn't work

Compiling the C++ model
starting worker pid=14386 on localhost:11575 at 13:39:28.187
starting worker pid=14412 on localhost:11575 at 13:39:28.570

SAMPLING FOR MODEL 'gaussian(identity) brms-model' NOW (CHAIN 1).
  [1] "Error : Rejecting initial value:"                                                                     
  [2] "  Log probability evaluates to log(0), i.e. negative infinity."                                       
  [3] "  Stan can't start sampling from this initial value."                                                 
  [4] "Rejecting initial value:"                                                                             
  [5] "  Log probability evaluates to log(0), i.e. negative infinity."                                       
  [6] "  Stan can't start sampling from this initial value."                                                                                                                       
...                                    
[300] "  Stan can't start sampling from this initial value."                                                 
[301] "Initialization between (--2, -2) failed after 100 attempts. "                                         
[302] " Try specifying initial values, reducing ranges of constrained values, or reparameterizing the model."
error occurred during calling the sampler; sampling not done
here are whatever error messages were returned
[[1]]
Stan model 'gaussian(identity) brms-model' does not contain samples.

[[2]]
Stan model 'gaussian(identity) brms-model' does not contain samples.

Warning message:
In .local(object, ...) :
  some chains had errors; consider specifying chains = 1 to debug
```

**Comment by joelgombin** [comment]
After looking at the generated STAN code, I think this may have to do with the centerised reparameterisation used for FE - I guess the constraints should be transformed accordingly.

**Comment by paul-buerkner** [comment]
The problem is that Stan trys to initilize values between -2 and 2 by default. But your prior only has mass between 2 and 4 and no gradient information elsewhere. Accordingly, initilialization must fail. 
You should consider using 

```
test2 <- brm(dist ~ speed, data = cars, family = gaussian(), 
             prior = set_prior("uniform(2, 4)", class = "b", coef = "speed"),
             init_r = 5)
```
to initilize between -5 and 5. This should work for your particular example (although you will get quite a few divergent transition warnings, because the sampler trys to go beyond the boundaries).

A word of caution though: In Stan it is generalliy not a good idea to but hard bounded priors (such as the uniform distribution) on unbounded parameters. If you are using those priors it would be better to restrict your parameter to the desired values, say, `vector<lower=2, upper=4>[K] b`, This is currently not possible to do this automatically in `brms` and since you may put those priors only on a subset of `b` it wouldn't even be possible without splitting up `b` in multiple parts.

**Comment by joelgombin** [comment]
Thanks for the answer. I am aware of the Stan authors' warnings regarding the uniform distribution, but in my use case I do have hard constraints (the beta coefficients are proportions). Restricting the parameters is obviously a better approach (that I've used when modelling directly through Stan), but as you said that's not possible in `brms`. I was just testing `brms` to see if it could be fit for my usecase :) Thanks anyway for the explanation.

**Comment by paul-buerkner** [comment]
Maybe it would be possible, but only if you put the prior on the whole FE vector. Then, I could just read out the boundaries and write them into the Stan code. Could you provide an example for how one of your real models would look like if specified in `brms`. Then I could tell you if I can implement it.

**Comment by joelgombin** [comment]
yep, that would be something like:

```r
brm(Y ~ 0 | trunc(lb = 0, ub = 1) + x1 + x2 + x3 + x4, data = data, family = gaussian(link = "identity"), prior = set_prior("uniform(0, 1)", class = "b"))
```

**Comment by joelgombin** [comment]
Another possibility would be to use gamma priors...

**Comment by paul-buerkner** [comment]
Thanks, I guess I can put this in, since you use the same restrictions for the whole fixed effects vector.

**Comment by joelgombin** [comment]
Yep. Beyond that I guess I can just adapt the generated Stan code ;-)

**Comment by paul-buerkner** [comment]
I agree, but I think it is still worth implementing it. :-)

**Comment by paul-buerkner** [comment]
I have just pushed some commits to github allowing to restrict parameters. For now this is only allowed for the FE parameter vector `b`, but more may follow. Using your test example you can now write:

```R
test1 <- brm(dist ~ speed, data = cars, family = gaussian(), 
             prior = set_prior("uniform(2,4)", class = "b", lb = 2, ub = 4))
``` 

This will ensure that `b` is bounded between 2 and 4. You can of course use other values for `lb` and `ub` or specify only one of them. For instance, you can also do

```R
test2 <- brm(dist ~ speed, data = cars, family = gaussian(), 
             prior = set_prior("normal(0,10)", class = "b", lb = 0))
```
to define a half-normal prior on `b`. I am still a bit sceptical about the range of application of bounded fixed effects parameters, but since it doesn't interfere with other functionality (and was relatively easy to implement), I think it's good to have it in.

Would you mind trying it with one of your models and give me some feedback whether it also works for you? You can download the latest version from github using 
```R
devtools::install_github("paul-buerkner/brms")
```

**Comment by joelgombin** [comment]
Thanks. I'm not completely sure that works, though. For example, try:

```r
brm(dist ~ speed, data = cars, family = gaussian(link = "identity"), prior = set_prior("normal(0, 10)", class = "b", lb = 2, ub = 4), init_r = 5)
```

This gives you:

```r
 Family: gaussian (identity) 
Formula: dist ~ speed 
   Data: cars (Number of observations: 50) 
Samples: 2 chains, each with iter = 2000; warmup = 500; thin = 1; 
         total post-warmup samples = 3000
   WAIC: 419.34
 
Fixed Effects: 
          Estimate Est.Error l-95% CI u-95% CI Eff.Sample Rhat
Intercept   -12.92      4.67   -20.25    -1.67       1969    1
speed         3.62      0.27     2.98     3.98       2016    1
```

So it seems the boundaries didn't apply to the intercept.
However, if you do:

```r
brm(dist ~ speed, data = cars, family = gaussian(link = "identity"), prior = set_prior("normal(0, 10)", class = "b", lb = 4, ub = 6), init_r = 5)
```
you get:
```r
 Family: gaussian (identity) 
Formula: dist ~ speed 
   Data: cars (Number of observations: 50) 
Samples: 2 chains, each with iter = 2000; warmup = 500; thin = 1; 
         total post-warmup samples = 3000
   WAIC: 419.17
 
Fixed Effects: 
          Estimate Est.Error l-95% CI u-95% CI Eff.Sample Rhat
Intercept   -23.42      4.33   -33.56   -16.50       1786    1
speed         4.31      0.24     4.01     4.91       1890    1
```

so the issue seems to be only with the intercept, not the other b coefficients. And when I run my own models which don't have an intercept (`Y ~ 0 + x1 + x2 + x3` etc.), the sampler doesn't manage to initialise itself:

```r
 [2] "Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:"                        
  [3] "Exception thrown at line 23: stan::math::uniform_log: Upper bound parameter is 0.1, but must be greater than 0.1"                      
  [4] "If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,"
  [5] "but if this warning occurs often then your model may be either severely ill-conditioned or misspecified."                              
  [6] "Rejecting initial value:"                                                                                                              
  [7] "  Error evaluating the log probability at the initial value."        
```
So I can only guess here but I suspect it might be related to the reparameterisation of the models?

**Comment by joelgombin** [comment]
Ah, sorry, I hadn't seen the title of that commit: https://github.com/paul-buerkner/brms/commit/b7f68e330602818f65ef0893cbe483b906641111

**Comment by joelgombin** [comment]
Yes, that seems to work great (except for the fact that the intercept is not subjected to the bounds, but I understand for the time being that's intentionnal). Thanks!

**Comment by paul-buerkner** [comment]
Thanks for testing. You are right that the intercept is currently not restricted, I should have mentioned this when asking you to retest it. The reason is that even if we would apply the bounds on the temporary intercept, the real intercept would be bounded somewhere else, also long as some predictors have non-zero mean. 

From the user side you can easily put the bounds also on the intercept (assuming you don't have factors as predictors) by defining a new variable in your dataset (e.g., `mydata$intercept <- 1` ) and then remove the default intercept from the model:

```R
cars$intercept <- 1
test1 <- brm(dist ~ -1 + intercept + speed, data = cars, family = gaussian(), 
             prior = set_prior("uniform(2,4)", class = "b", lb = 2, ub = 4))
```

In fact, making `intercept` a reserved variable name and doing the `intercept <- 1` thing internally, might generally help people who want to set a prior on the intercept directly as compared to just setting the prior to the temporary one and I could make this easily work even with factors. I will have to think about it a bit more though and any opinion from the user's perspective is much appreciated.

**Comment by joelgombin** [comment]
Yeah, I didn't think about that (I didn't really have to as my models don't have an intercept). I think maybe it's more flexible to let the user do this himself, provided it's clearly explained in the documentation? Nice trick anyway.

**Comment by paul-buerkner** [comment]
It is now possible to use the reserved (and internally generated) variable `intercept` to also restrict the fixed effects intercept if desired. More importantly, this will allow to define priors on the real intercept not on the temporary one that is used for the centered reparameterization. I decided to generate the variable internally as it won't work reliably with factors, otherwise.

I also documented any changes and new features so I guess this issue can be closed. Thank you very much for your ideas and your help! :-)
